calicoctl/commands/node/checksystem.go

```diff
@@ -0,0 +1,220 @@
+	if !goversion.CompareNormalized(kernelVersionStr, minKernelVersion, ">=") {
+
+		// Prints "FAIL" if current version is not >= minimum required version
+		fmt.Printf("\t\t\t\t\tFAIL\n")
+		log.Printf("Minimum kernel version to run Calico is %s. Detected kernel version: %s", minKernelVersion, string(kernelVersion))
+		return errors.New("Kernel version mismatch")
+	}
+
+	// Prints "OK" if current version is >= minimum required version
+	fmt.Printf("\t\t\t\t\tOK\n")
+
+	return nil
+}
+
+// checkKernelModules checks for all the required kernel modules in the system
+func checkKernelModules() error {
+
+	kernelVersion, err := exec.Command("uname", "-r").Output()
+	if err != nil {
+		log.Println(err)
+	}
+
+	// To strip the trailing `\n`
+	kernelVersionStr := strings.TrimSpace(string(kernelVersion))
+
+	// File path to Loadable kernel modules
+	modulesLoadablePath := fmt.Sprintf("/lib/modules/%s/modules.dep", kernelVersionStr)
+
+	// File path to Builtin kernel modules
+	modulesBuiltinPath := fmt.Sprintf("/lib/modules/%s/modules.builtin", kernelVersionStr)
+
+	// Keep track of modules that are not found
+	modulesNotFound := []string{}
+
+	// Execute lsmod and cache the result
+	lsmodOut, err := exec.Command("lsmod").Output()
+	if err != nil {
+		log.Println(err)
+	}
+
+	// Go through all the required modules and check Loadable and Builtin in order
+	for _, v := range requiredModules {
+		err = checkModule(modulesLoadablePath, v, kernelVersionStr)
+
+		// Check Builtin modules if not found in Loadable
+		if err != nil {
+			err = checkModule(modulesBuiltinPath, v, kernelVersionStr)
+
+			// Check if it's in lsmod, if not found in Builtin either
+			if err != nil {
+
+				regex, err := regexp.Compile(v)
+				if err != nil {
+					log.Fatalf("Error: %v\n", err)
+				}
+
+				if !regex.MatchString(string(lsmodOut)) {
+					log.Printf("WARNING: Unable to detect the %s module as Loaded/Builtin module or lsmod\n", v)
+					modulesNotFound = append(modulesNotFound, v)
+					fmt.Printf("\t\t%-20s%s", v, "\t\t\t\t\tFAIL\n")
+				} else {
+					fmt.Printf("\t\t%-20s%s", v, "\t\t\t\t\tOK\n")
+				}
+			} else {
+				fmt.Printf("\t\t%-20s%s", v, "\t\t\t\t\tOK\n")
+			}
+		} else {
+			fmt.Printf("\t\t%-20s%s", v, "\t\t\t\t\tOK\n")
+		}
+	}
+
+	// If there are still any modules not found then return an error
+	if len(modulesNotFound) > 0 {
+
+		// ip6_tables is not a required module for ipv4 setups, so just print
+		// a warning instead of failing the systemc heck
+		if len(modulesNotFound) == 1 && modulesNotFound[0] == "ip6_tables" {
+			fmt.Printf("WARNING: Unable to detect ip6_tables kernel module")
+			return nil
+		}
+		return errors.New("One of more kernel modules missing")
+	}
+
+	return nil
+}
+
+// checkModule is a utility function used by `checkKernelModules`
+// it opens the file provided and checks if the module passed in
+// as an argument exists for the provided kernelVersion
+func checkModule(filename, module, kernelVersion string) error {
```

Comment:
I still don't see the point of having a page of code to replace a single OS command, but if you really want it (and the unit testing etc. responsibilities that will go with this), then fine.